rmw_fastrtps_shared_cpp/src/custom_publisher_info.cpp

```diff
@@ -358,14 +360,16 @@ void RMWPublisherEvent::update_matched(
 {
   std::lock_guard<std::mutex> lock(on_new_event_m_);
 
-  matched_status_.total_count = total_count;
-  matched_status_.total_count_change += total_count_change;
-  matched_status_.current_count = current_count;
-  matched_status_.current_count_change += current_count_change;
+  if (on_new_event_cb_[RMW_EVENT_PUBLICATION_MATCHED]) {
```

Comment:
If we set the callback via `set_on_new_qos_event_callback`, the same repeated problem can be observed? since it will eventually calls `trigger_event`?

Reply:
I think this would be the expected behavior. When the user sets a callback (for instance, for the event executor), we should first notify the user callback, and then notify the event on any pending `rmw_wait` call.

One thing to notice, if `rmw_take_event` is called within the execution of the callback, the corresponding guard condition trigger value is changed to false [here](https://github.com/ros2/rmw_fastrtps/blob/5913bd46155fdb1c384896f3aa6823d7d2e44301/rmw_fastrtps_shared_cpp/src/custom_publisher_info.cpp#L197)

Reply:
@MiguelCompany thanks for the explanation!